### DISCUSSION TITLE
cpp-eval-command-discord

### DISCUSSION THREAD

**Discussion by giscus** [brainboxdotcc/dpp-comments#46] [General]
# cpp-eval-command-discord

A lightweight C++ Discord API library supporting the entire Discord API, including Slash Commands, Voice/Audio, Sharding, Clustering and more!

https://dpp.dev/cpp-eval-command-discord.html

**Comment by joshcoyne** [comment]
Hi! Is this a good tutorial to have? While at a high level it emulates javascript eval(), javascript is fully sandboxed. This however is not. I know you have a NOTE, but I think you'd need to accomplish something that avoids a red NOTE to offer a solution like this in the first place.

**Reply by Mishura4** [reply]
There is no such thing without an elaborate script manager using an external library.

It's just something people ask how to do for some reason so we added it to the docs, but yes, nobody in their right mind should have an eval command in a production bot.

**Reply by joshcoyne** [reply]
I agree, I feel the overhead of making a "real" C++ sandbox would require understanding how to virtualize/sandbox your environment so the binary can't call into the host environment running DPP. And that's not a scope you wanna provide in the docs.

I won't press on the issue further, I just wanted to comment on it as a moral judgement.

I love dpp and have built a really cool bot with it so keep up the great work :)

**Reply by braindigitalis** [reply]
javascript isn't sandboxed unless youre running it in a browser - eval() in js is just as dangerous as this example.
It's not something you should ever do. One of my bots integrates libduktape to properly sandbox javascript for end users to run, this is essentially the only safe approach, native code can't ever be made safe (even if you virtualise it, someone might break out of the virtual environment via a hypervisor exploit)

**Reply by joshcoyne** [reply]
I disagree that libduktape eval is just as bad as c++ eval, unless you're providing sys i/o or escalation callbacks, which aren't exposed out of the box for most javascript engines (excepting system ones like node)

Is it possible for someone to find a duktape exploit and then write some memory into your system? Sure. But that's alot harder than including <iostream> in a raw file and compiling it, not to mention how trivial it is to get system details w native code that could be used as an attack vector. Figuring out your kernel version and reading CVEs is not much harder than including iostream :)

**Reply by Jaskowicz1** [reply]
I think Brain meant that javascript isn't sandboxed so he uses libduktape to sandbox javascript (basically, he's saying the only way to pretty much be safe with eval in JS is to use libduktape, defending libduktape).

**Reply by joshcoyne** [reply]
My understanding is javascript interpreters out the box (v8/duktape/tracemonkey/etc) have no ability for js code to escalate to syscalls without binding the native code calls in, like nodes file io package for instance. 

Obvs exploits can happen but we have responsibility as programmers in being proactive in our architecture design. If developers or users ask for something that can easily make their system insecure or broken, I believe the right course of action is to educate on why that is not a valid solution. Especially if folks are coming from interpreter environments such as js/python into c++.

**Reply by Jaskowicz1** [reply]
That's what brain is saying :P

**Reply by joshcoyne** [reply]
Then why is a horrible security solution being provided instead of said education, if we are saying the same things? Putting a warning at the top isn't the same thing, it's still an official example.

**Reply by Jaskowicz1** [reply]
Because we don't agree with eval stuff, we don't want to support people making them, but people kept asking (like @Mishura4 said) so we provided a doc page.

None of us as volunteers have any care to support eval and improve it, so we won't. However, if you wish to improve it yourself, we'd gladly accept a PR!

But overall, we would rather not have any eval stuff (quite literally because this conversation, we do not want to be responsible for any issues and people WILL come into the Discord and have a go if we don't declare something like that, luring people in by trying to be secure is wrong and can lead to many many many issues on our end).

**Reply by Mishura4** [reply]
> Then why is a horrible security solution being provided instead of said education, if we are saying the same things? Putting a warning at the top isn't the same thing, it's still an official example.

Because people really don't like being told "don't do that"

**Comment by Neko-Life** [comment]
Do we need to remove this or what? What do you want honestly?

It is explicitly stated:
> Needless to say, this is very dangerous. If you are asking how to do this, and want to put this into your bot, we trust that you have a very good reason to do so and have considered alternatives before resorting to this. The code below is for educational purposes only

Any sane C++ developer won't be doing this in their production bot, even in JS it is really not recommended to use eval outside experiment/debugging purpose
